File: MyMonitorHub.JavaAgent/src/main/java/mymonitorhub/agent/core/CommandProcessingThread.java
package mymonitorhub.agent.core;

import mymonitorhub.agent.model.EventType;
import mymonitorhub.agent.model.ItemStyle;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.Socket;
import java.nio.charset.Charset;
import java.util.Date;

/**
 * Handles a single inbound MCM connection.
 * Supports commands: fireEvent, ping.
 * Mirrors C# CommandProcessingThread.
 */
class CommandProcessingThread {

    private static final Logger log = LoggerFactory.getLogger(CommandProcessingThread.class);

    private final Socket socket;
    private McmProtocol mcmProtocol;

    CommandProcessingThread(Socket socket) {
        this.socket = socket;
    }

    void start() {
        Thread t = new Thread(new Runnable() {
            @Override
            public void run() {
                doRun();
            }
        });
        t.setName("cmd-processor");
        t.setDaemon(true);
        t.start();
    }

    private void doRun() {
        try {
            socket.setSoTimeout(30000);
            mcmProtocol = new McmProtocol(socket);
            mcmProtocol.read();

            String data = mcmProtocol.getHeader();
            String command = getData(data, "Command: ");
            if (command == null) {
                sayFailed("Missing 'Command' parameter");
                return;
            }

            switch (command) {
                case "fireEvent":
                    fireEventCommand(data);
                    break;
                case "ping":
                    sayPassed();
                    break;
                default:
                    sayFailed("Unrecognized command: " + command);
            }
        } catch (Exception e) {
            log.error("run> {}\n{}", e.getMessage(), e.getStackTrace());
        } finally {
            try { socket.close(); } catch (IOException ignored) {}
        }
    }

    private void fireEventCommand(String data) throws IOException {
        String group    = getData(data, "Group: ");
        String category = getData(data, "Category: ");
        String name     = getData(data, "Name: ");
        String desc     = getData(data, "Description: ");
        String status   = getData(data, "Status: ");
        String style    = getData(data, "Style: ");

        if (group == null || category == null || name == null
                || desc == null || status == null || style == null) {
            sayFailed("Missing a parameter");
            return;
        }

        EventType stat;
        switch (status.toUpperCase()) {
            case "OK":   stat = EventType.Ok;   break;
            case "FAIL": stat = EventType.Fail; break;
            default:
                sayFailed("Invalid status code.");
                return;
        }

        ItemStyle styleEnum;
        switch (style.toUpperCase()) {
            case "NO_HEALTH":    styleEnum = ItemStyle.NoHealth;    break;
            case "SENDS_HEALTH": styleEnum = ItemStyle.SendsHealth; break;
            default:
                sayFailed("Invalid style code.");
                return;
        }

        Monitor.fireEvent(group, category, name, desc, stat, styleEnum, new Date());
        sayPassed();
    }

    private static String getData(String data, String searchFor) {
        String marker = "\r\n" + searchFor;
        int start = data.indexOf(marker);
        if (start == -1) return null;
        int valueStart = start + marker.length();
        int end = data.indexOf("\r\n", valueStart);
        if (end == -1) return null;
        return data.substring(valueStart, end);
    }

    private void sayPassed() throws IOException {
        mcmProtocol.write("Status: OK\r\n", null);
    }

    private void sayFailed(String reason) throws IOException {
        mcmProtocol.write("Status: Fail\r\n", reason.getBytes(Charset.forName("UTF-8")));
    }
}

File: MyMonitorHub.Agent/MyMonitorHub.Agent.Core/Core/CommandProcessingThread.cs
using System;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using MyMonitorHub.Common.WebApi;
using MyMonitorHub.Agent.Client;
using Serilog;

namespace MyMonitorHub.Agent.Core
{
    /// <summary>
    ///     This process handles the incomming event which will basically put it in the queue. We batch up events and send
    ///     them every minute rather than immediately.
    /// </summary>
    internal sealed class CommandProcessingThread
    {
        private static readonly ILogger Logger = Log.ForContext<CommandProcessingThread>();
        private readonly Socket _socket;
        private McmProtocol? _mcmProtocol;

        internal CommandProcessingThread(Socket socket)
        {
            _socket = socket;
        }

        // constructor

        /// <summary>
        ///     Main starting point of thread.
        /// </summary>
        private void Run()
        {
            try
            {
                using (_socket)
                {
                    try
                    {
                        _socket.ReceiveTimeout = 30000; // 30 seconds

                        _mcmProtocol = new McmProtocol(_socket);

                        // the body currently isn't used
                        _mcmProtocol.Read();

                        // the header has all my params in it.
                        var data = _mcmProtocol.Header;

                        // I must have got the full response now
                        var command = GetData(data, "Command: ");
                        if (command == null)
                        {
                            SayFailed("Missing 'Command' parameter");
                            return;
                        }
                        if (command == "fireEvent")
                        {
                            FireEventCommand(data);
                        }
                        else if (command == "ping")
                        {
                            Ping();
                        }
                        else
                        {
                            SayFailed("Unrecognized command: " + command);
                        }
                    }
                    catch (Exception e)
                    {
                        Logger.Error($"run> {e.Message}\r\n{e.StackTrace}");
                    }
                } // using _socket

            }
            catch (Exception ex)
            {
                // these seems extremely unlikely but I will trap for it anyways
                Logger.Error($"run> dispose problem> {ex.Message}\r\n{ex.StackTrace}");
            }
        }

        // run

        private void Ping()
        {
            SayPassed();
        }

        private void FireEventCommand(string data)
        {
            var group = GetData(data, "Group: ");
            var category = GetData(data, "Category: ");
            var name = GetData(data, "Name: ");
            var desc = GetData(data, "Description: ");
            var status = GetData(data, "Status: ");
            var style = GetData(data, "Style: ");
            if (group == null || category == null || name == null || desc == null || status == null || style == null)
            {
                SayFailed("Missing a parameter");
                return;
            }
            EventType stat;
            switch (status.ToUpper())
            {
                case "OK":
                    stat = EventType.Ok;
                    break;
                //case "WARN":
                //    stat = Event.StatusWarn;
                //    break;
                case "FAIL":
                    stat = EventType.Fail;
                    break;
                default:
                    SayFailed("Invalid status code.");
                    return;
            }

            ItemStyle styleInt;
            if (style.ToUpper() == "NO_HEALTH")
            {
                styleInt = ItemStyle.NoHealth;
            }
            else if (style.ToUpper() == "SENDS_HEALTH")
            {
                styleInt = ItemStyle.SendsHealth;
            }
            else
            {
                SayFailed("Invalid style code.");
                return;
            }

            Monitor.FireEvent(group, category, name, desc, stat, styleInt, DateTime.Now);
            SayPassed();
        }


        private static string? GetData(string data, string searchFor)
        {
            var start = data.IndexOf("\r\n" + searchFor, StringComparison.Ordinal);
            if (start == -1)
            {
                return null;
            }
            var endIndex = data.IndexOf("\r\n", start + 2, StringComparison.Ordinal);

            return data.Substring(start + searchFor.Length + 2, endIndex - (start + searchFor.Length + 2));
        }

        private void SayFailed(string reason)
        {
            try
            {
                // if what was read was encrypted then the response should also be encrypted
                _mcmProtocol!.Write("Status: Fail\r\n", Encoding.UTF8.GetBytes(reason));
            }
            catch (IOException e)
            {
                Logger.Error(e, "Failed to say Fail");
            }
        }

        private void SayPassed()
        {
            try
            {
                // if what was read was encrypted then the response should also be encrypted
                _mcmProtocol!.Write("Status: OK\r\n", null);
            }
            catch (IOException e)
            {
                Logger.Error(e, "Failed to say Pass");
            }
        }

        internal void Start()
        {
            var thread = new Thread(Run);
            thread.Start();
        }

    }

    // class
}